src/core/ir/discriminator.rs

```diff
@@ -33,1212 +80,13 @@ where
     }
 
     fn set_type_name(&'json mut self, type_name: String) -> Result<(), Self::Error> {
-        if let Some(obj) = self.as_object_mut() {
+        if self.is_null() {
+            Ok(())
+        } else if let Some(obj) = self.as_object_mut() {
             obj.insert_key(TYPENAME_FIELD, T::string(type_name.into()));
-
             Ok(())
         } else {
-            bail!("Expected object")
-        }
-    }
-}
-
-/// Resolver for type member of a union.
-/// Based on type definitions and the provided value, it can
-/// resolve the type of the value.
-///
-/// ## Resolution algorithm
-///
-/// The resolution algorithm is based on the following points:
-/// - The common set of fields is the set of all fields that are defined in the
-///   type members of the union.
-/// - If the resolved value is a list, then the resolution should be run for
-///   every entry in the list as a separate value.
-/// - If a field from the common set is present in the resolved value, then the
-///   result type is one of the types that have this field.
-/// - If a field from the common set is required in some types and this field is
-///   not present in the resolved value, then the result type is not one of
-///   those types.
-/// - By repeating the checks from above for every field in the common set, we
-///   will end up with a smaller set of possible types and, more likely, with
-///   only a single possible type.
-
-#[derive(Clone)]
-pub struct Discriminator {
```

Comment:
if you ask the LLM to generate simple schema with union it will generate something like following:

```graphql
union SearchResult = Book | Author

type Book {
  id: ID!
  title: String!
  author: String!
}

type Author {
  id: ID!
  name: String!
  books: [Book!]
}

type Query {
  search(term: String!): SearchResult
}
```

I don't think API in that case will generate "proper" enums layout to support the discriminator change